### PR TITLE
net: Countermeasures against eclipse attacks

### PR DESCRIPTION
**Pull request by Pythonix**
A few years ago, the Eclipse attack against the Bitcoin network was published. This PR represents a collection of the commits that were implemented upstream as a fix. For reference please see https://github.com/bitcoin/bitcoin/pull/5941

> This implements countermeasures 1, 2, and 6 from the http://cs-people.bu.edu/heilman/eclipse/ paper.

### CODE REVIEW

File: src/addrman.cpp [hunk 1]
```diff
@@ -278,48 +212,47 @@ void CAddrMan::Good_(const CService &addr, int64_t nTime)
...
     }
 
     // if no bucket is found, something bad happened;
     // TODO: maybe re-add the node, but for now, just bail out
-    if (nUBucket == -1) return;
+    if (nUBucket == -1)
+        return;
 
-    LogPrint(BCLog::LogFlags::ADDRMAN, "Moving %s to tried", addr.ToString());
+    LogPrint("addrman", "Moving %s to tried\n", addr.ToString());
```
Comment (div72) [inline src/addrman.cpp]: Could you revert this line?

File: src/addrman.h [hunk 2]
```diff
@@ -319,83 +327,143 @@ class CAddrMan
...
+            if (it->second.fInTried == false && it->second.nRefCount == 0) {
+                std::map<int, CAddrInfo>::const_iterator itCopy = it++;
+                Delete(itCopy->first);
+                nLostUnk++;
+            } else {
+                it++;
+            }
+        }
+        if (nLost + nLostUnk > 0) {
+            LogPrint("addrman", "addrman lost %i new and %i tried addresses due to collisions\n", nLostUnk, nLost);
```
Comment (div72) [inline src/addrman.h]: Ditto.

File: src/addrman.cpp (hunk 1 above)
Comment (Pythonix) [reply src/addrman.cpp]: Good catch, thank you.

**Comment by jamescowens** [comment]
@div72 @Pythonix What do we want to do with these changes. How much jogging from old upstream do we want to do to fix vulnerabilities versus ripping the net code out and replacing with current upstream. Also @div72 what is your opinion on merging this and the other net PR's in the interim?

Also, let's try to be better at catching regressions like the above pointed out by @div72 where an upstream module has already been replaced (like the logger) and certain statements are already ahead of a patch like this.

File: src/addrman.h [hunk 3]
```diff
@@ -415,7 +415,7 @@ class CAddrMan
             }
         }
         if (nLost + nLostUnk > 0) {
-            LogPrint("addrman", "addrman lost %i new and %i tried addresses due to collisions\n", nLostUnk, nLost);
```
Comment (jamescowens) [inline src/addrman.h]: We don't put the \n explicitly like Bitcoin, rather it is in the logger itself. Please remove the \n as well. Thanks.

**Comment by div72** [comment]
@jamescowens Lets merge this and the other PR for now. Although I support "ripping the net code out and replacing with current upstream" approach more, PRs like this should make the diff lesser and make the bigger PRs easier to review. And these two PRs are enough self-contained that no issues should popup.

**Comment by jamescowens** [comment]
agreed.

File: src/addrman.h [hunk 4]
```diff
@@ -319,83 +327,143 @@ class CAddrMan
...
+            if (it->second.fInTried == false && it->second.nRefCount == 0) {
+                std::map<int, CAddrInfo>::const_iterator itCopy = it++;
+                Delete(itCopy->first);
+                nLostUnk++;
+            } else {
+                it++;
+            }
+        }
+        if (nLost + nLostUnk > 0) {
+            LogPrint(BCLog::LogFlags::ADDRMAN, "addrman lost %i new and %i tried addresses due to collisions\n", nLostUnk, nLost);
```
Comment (jamescowens) [inline src/addrman.h]:
```suggestion
            LogPrint(BCLog::LogFlags::ADDRMAN, "addrman lost %i new and %i tried addresses due to collisions", nLostUnk, nLost);
```

**Review by jamescowens — APPROVED** [review]
I went ahead and just made the small edit to remove the \n from the LogPrint. All else was in order.

**Pull request merged**